src/session.c

```diff
@@ -178,56 +180,129 @@ static P11PROV_SESSION *p11prov_session_ref(P11PROV_SESSION *session)
     return NULL;
 }
 
-static CK_RV p11prov_session_open(P11PROV_SESSION *session, bool login,
-                                  CK_UTF8CHAR_PTR pin, CK_ULONG pinlen)
+/* Calculates the start time and then nano-sleeps by 'interval' time.
+ * On the first invocation the content of start_time must be 0.
+ * The content of start_time must not be altered outside this function after
+ * the first invocation.
+ * This function does not guarantee each sleep is 'interval' long.
+ *
+ * Returns true if max_wait has not been reached yet.
+ * Returns false on an error or if max_wait is exceeded.
+ */
+static bool cyclewait_with_timeout(uint64_t max_wait, uint64_t interval,
+                                   uint64_t *start_time)
+{
+#define NANOS_SEC 1000000000
+    uint64_t current_time;
+    struct timespec ts;
+    int ret;
+
+    ret = clock_gettime(CLOCK_MONOTONIC, &ts);
+    if (ret != 0) {
+        return false;
+    }
+
+    current_time = ts.tv_sec * NANOS_SEC + ts.tv_nsec;
+    if (*start_time == 0) {
+        *start_time = current_time;
+    } else {
+        if (current_time > *start_time + max_wait) {
+            return false;
+        }
+    }
+
+    ts.tv_sec = interval / NANOS_SEC;
+    ts.tv_nsec = interval % NANOS_SEC;
+    ret = nanosleep(&ts, NULL);
+    if (ret != 0 && ret != EINTR) {
+        return false;
+    }
+
+    return true;
+}
+
+/* in nanoseconds, 1 seconds */
+#define MAX_WAIT 1000000000
+/* sleep interval, 50 microseconds (max 20 attempts) */
+#define SLEEP 50000
+static CK_RV internal_session_open(P11PROV_SESSION_POOL *p, CK_SLOT_ID slot,
+                                   CK_SESSION_HANDLE *session)
 {
-    P11PROV_SESSION_POOL *p = session->pool;
-    CK_ULONG cur_sessions;
     CK_FUNCTION_LIST *f;
+    bool wait_ok = true;
+    CK_ULONG cs = 0;
+    uint64_t startime = 0;
     CK_RV ret;
 
-    f = p11prov_ctx_fns(session->provctx);
+    f = p11prov_ctx_fns(p->provctx);
     if (f == NULL) {
         return CKR_FUNCTION_FAILED;
     }
 
-    cur_sessions = __atomic_add_fetch(&p->cur_sessions, 1, __ATOMIC_SEQ_CST);
-    if (cur_sessions > p->max_sessions) {
-        ret = CKR_SESSION_COUNT;
-        P11PROV_raise(session->provctx, ret, "Max sessions limit");
-        goto done;
+    while (wait_ok) {
+        if (cs == 0) {
+            cs = __atomic_add_fetch(&p->cur_sessions, 1, __ATOMIC_SEQ_CST);
+            if (cs > p->max_sessions) {
+                (void)__atomic_sub_fetch(&p->cur_sessions, 1, __ATOMIC_SEQ_CST);
+                ret = CKR_SESSION_COUNT;
+                cs = 0;
+                wait_ok = cyclewait_with_timeout(MAX_WAIT, SLEEP, &startime);
+                continue;
+            }
+        }
+
+        wait_ok = false;
+        ret = f->C_OpenSession(slot, CKF_SERIAL_SESSION, NULL, NULL, session);
+        if (ret == CKR_SESSION_COUNT) {
```

Comment:
Should this also adjust the session pool limits if maximum was reached?
Or is this intentionally left not adjusting it, in case some other library is using the same token?

Reply:
Yeah I do not want to unnecessarily reduce maximums, because the token may be used concurrently by multiple applications at some point in time so the failure may be spurious.
Thinking of a network HSM that have some quota on concurrent sessions based on source IP or something, and similar other HW limitations for directly attached HSMs.

We should never really hit this, because we also have a soft limit to try to avoid to hog all available sessions anyway.